### PR TITLE
Feature test usart into release 

### PR DESCRIPTION
**Pull request by PatrickDekker98**

### CODE REVIEW

File: test/Makefile [hunk 1]
```diff
@@ -9,10 +9,12 @@
 #############################################################################
 
 # source files in this project (main.cpp is automatically assumed)
-SOURCES := $(wildcard ../code/src/*.cpp)
+#SOURCES := $(wildcard ../code/src/*.cpp)
```
Comment (itzandroidtab) [inline test/Makefile]: comments?

File: code/headers/test_usart.hpp [hunk 2]
```diff
@@ -2,20 +2,21 @@
...
-        queue_c<uint8_t, 250> input_buffer;
+        queue_c<uint8_t, 250> receive_buffer;
 
 
     public:
 
-        test_usart_c(unsigned int baudrate, uart_ports_c usart_port);
+        test_usart_c();
 
         /// @brief does not actualy disable anything
```
Comment (itzandroidtab) [inline code/headers/test_usart.hpp]: this comment is wrong btw. The function name is enable

File: code/headers/test_usart.hpp [hunk 3]
```diff
@@ -47,5 +48,17 @@ namespace r2d2 {
         ///@return unsigned int always 1
```
Comment (itzandroidtab) [inline code/headers/test_usart.hpp]: this doesn't always return 1 anymore

File: code/main.cpp [hunk 4]
```diff
@@ -8,5 +8,6 @@ int main(void) {
     WDT->WDT_MR = WDT_MR_WDDIS;
     hwlib::wait_ms(1000);
     hwlib::cout << "this works on arduino";
+    auto hwusart = r2d2::hardware_usart_c(9600, r2d2::uart_ports_c::uart1);
```
Comment (itzandroidtab) [inline code/main.cpp:11]: maybe we need to move all the usart to a usart namespace

File: code/src/test_usart.cpp [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+
+    void test_usart_c::set_receive_string(const std::string &str) {
+        receive_buffer.clear();
+        for(char c : str){
+            set_receive_byte(c);
+        }
+    }
+
+    void test_usart_c::set_receive_bytes(const std::vector<uint8_t> &bytes){
+        receive_buffer.clear();
```
Comment (itzandroidtab) [inline code/src/test_usart.cpp]: why are you clearing the receive_buffer? what if i want to add some bytes first with the set_receive_byte function

File: code/src/test_usart.cpp [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+        for(auto &byte : bytes){
+            set_receive_byte(byte);
+        }
+    }
+
+    void test_usart_c::set_receive_byte(const uint8_t byte){
```
Comment (itzandroidtab) [inline code/src/test_usart.cpp]: this function sounds like the byte i set here will be the only byte that will be in the receive_buffer. Maybe a name change to something like ```add_receive_byte()``` is better here

Comment (itzandroidtab) [reply code/src/test_usart.cpp]: or a clear before the push is oke as well i guess.

File: code/src/test_usart.cpp [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+    void test_usart_c::putc(char c) {
+        send(c);
+    }
+
+    uint8_t test_usart_c::receive() {
+        return receive_buffer.copy_and_pop();
+    }
+
+    bool test_usart_c::char_available() {
```
Comment (itzandroidtab) [inline code/src/test_usart.cpp:28]: What is the difference between ```char_available()``` and ```available()```? seems a bit redundant they return the exact same atm

File: test/main.cpp [hunk 8]
```diff
@@ -1,4 +1,36 @@
 #include "ostream"
-
 #define CATCH_CONFIG_MAIN
 #include <catch.hpp>
+
+#include <test_usart.hpp>
+#include <uart_ports.hpp>
+
+TEST_CASE("test_usart_c sends","[test_usart_c]"){
```
Comment (itzandroidtab) [inline test/main.cpp:9]: maybe a test with the ```set_receive_string()``` and the ```set_receive_bytes()```?

**Review by itzandroidtab — CHANGES_REQUESTED** [review]
Looks pretty nice to test the usart bus with actual data. Got some small points above.

File: code/main.cpp (hunk 4 above)
Comment (itzandroidtab) [reply code/main.cpp:11]: i made a branch with everything moved to their own namespace

File: code/src/test_usart.cpp (hunk 7 above)
Comment (PatrickDekker98) [reply code/src/test_usart.cpp:28]: char_available returns true if char is available, available returns the amound of bytes available

File: test/main.cpp (hunk 8 above)
Comment (PatrickDekker98) [reply test/main.cpp:9]: were added in https://github.com/R2D2-2019/USART_library/pull/12/commits/85fdce96b4074228fcd84ff4c1db1e03da94d3f2

File: code/src/test_usart.cpp (hunk 6 above)
Comment (PatrickDekker98) [reply code/src/test_usart.cpp]: was changed in https://github.com/R2D2-2019/USART_library/pull/12/commits/85fdce96b4074228fcd84ff4c1db1e03da94d3f2

File: code/headers/test_usart.hpp (hunk 3 above)
Comment (PatrickDekker98) [reply code/headers/test_usart.hpp]: edited in  https://github.com/R2D2-2019/USART_library/pull/12/commits/85fdce96b4074228fcd84ff4c1db1e03da94d3f2

File: code/headers/test_usart.hpp (hunk 2 above)
Comment (PatrickDekker98) [reply code/headers/test_usart.hpp]: edited in https://github.com/R2D2-2019/USART_library/pull/12/commits/85fdce96b4074228fcd84ff4c1db1e03da94d3f2

File: test/Makefile (hunk 1 above)
Comment (PatrickDekker98) [reply test/Makefile]: gone by https://github.com/R2D2-2019/USART_library/pull/12/commits/85fdce96b4074228fcd84ff4c1db1e03da94d3f2

File: code/src/test_usart.cpp (hunk 7 above)
Comment (itzandroidtab) [reply code/src/test_usart.cpp:28]: But both ```char_available()``` and ```available()``` return ```!receive_buffer.empty()``` that is unintended then i guess

**Review by itzandroidtab — CHANGES_REQUESTED** [review]
Looks good except the ```char_available()``` and the ```available()``` both return ```!receive_buffer.empty()``` and your comment suggests that ```available()``` should return a amount of data left

**Review by itzandroidtab — APPROVED** [review]

**Pull request merged**